src/software/ai/navigator/path_planner/hrvo/hrvo_simulator.cpp

```diff
@@ -0,0 +1,242 @@
+#include "hrvo_simulator.h"
+
+HRVOSimulator::HRVOSimulator() : robots(), world(std::nullopt), primitive_set() {}
+
+void HRVOSimulator::updateWorld(const World &world,
+                                const RobotConstants_t &robot_constants,
+                                Duration time_step)
+{
+    this->world = world;
+    // TODO (#2498): Update implementation to correctly support adding and removing agents
+    //               to represent the newly added and removed friendly/enemy robots in the
+    //               World.
+    robots.clear();
+    for (const Robot &friendly_robot : world.friendlyTeam().getAllRobots())
+    {
+        configureHRVORobot(friendly_robot, robot_constants, time_step);
+    }
+
+    for (const Robot &enemy_robot : world.enemyTeam().getAllRobots())
+    {
+        configureLVRobot(enemy_robot, robot_constants, time_step);
+    }
+}
+
+
+void HRVOSimulator::updatePrimitiveSet(const TbotsProto::PrimitiveSet &new_primitive_set,
+                                       Duration time_step)
+{
+    primitive_set = new_primitive_set;
+
+    // Update all friendly agent's primitives
+    for (auto &[robot_id, primitive] : primitive_set.robot_primitives())
+    {
+        auto friendly_robot = robots.find(robot_id);
+        if (friendly_robot == robots.end())
+        {
+            continue;
+        }
+        else
+        {
+            if (world.has_value())
+            {
+                friendly_robot->second->updatePrimitive(primitive, world.value(),
+                                                        time_step);
+            }
```

Comment:
nit: this if statement can wrap the for loop so it's not checked everytime (could also just return if it doesn't have a value, after caching the primitive_set